Add host address to BindTransportException message

mariaral opened the pull request:
When bind fails, show the host address in addition to the port. This
helps debugging cases with wrong "network.host" values.

Closes #48001

mariaral committed:
Add host address to BindTransportException message

When bind fails, show the host address in addition to the port. This
helps debugging cases with wrong "network.host" values.

Closes #48001

elasticmachine commented:
Pinging @elastic/es-distributed (:Distributed/Network)

DaveCTurner commented:
@elasticmachine test this please

original-brownbear commented on server/src/main/java/org/elasticsearch/http/AbstractHttpServerTransport.java:
```diff
@@ -174,7 +174,10 @@ private TransportAddress bindAddress(final InetAddress hostAddress) {
         });
         if (!success) {
-            throw new BindHttpException("Failed to bind to [" + port.getPortRangeString() + "]", lastException.get());
+            throw new BindHttpException(
```
This is failing forbidden APIs checks, see the logs:


```
Forbidden method invocation: java.net.InetAddress#getHostAddress() [use NetworkAddress format/formatAddress to print IP or IP+ports]
  in org.elasticsearch.transport.TcpTransport (TcpTransport.java:381)
Forbidden method invocation: java.net.InetAddress#getHostAddress() [use NetworkAddress format/formatAddress to print IP or IP+ports]
  in org.elasticsearch.http.AbstractHttpServerTransport (AbstractHttpServerTransport.java:178)
Scanned 5715 class file(s) for forbidden API invocations (in 5.20s), 2 error(s).
```

mariaral replied:
Hi @original-brownbear, thanks for reviewing this. I ran the tests but forgot to run the check task. Will fix this ASAP.

original-brownbear replied:
@mariaral FYI, you can just run these simple static code checks via:

```
./gradlew precommit 
```

no need to run a full `check` for those

mariaral replied:
Thanks @original-brownbear for the tip. I didn't know of the precommit task.

I'm thinking of extending `NetworkAddress.format()` to be able to handle PortsRange objects and use that instead of coping the same message in five places. I propose using the following format:
```
    IPv4 no port: 127.0.0.1
    IPv4 single port: 127.0.0.1:9300}
    IPv4 multiple ports: 127.0.0.1:[9300-9400]
    IPv6 multiple ports: [::1]:[9300-9400]
```
What do you think?

original-brownbear replied:
I only count two spots where we print the `PortsRange` but fine by me to keep it consistent via a utility method :)

mariaral replied:
I pushed a new commit. Please take a look.

mariaral committed:
Extend NetworkAddress.format() to handle port ranges

Extend NetworkAddress.format() to handle port ranges and use this
utility method instead of forbidden InetAddress.getHostAddress().

DaveCTurner commented:
@elasticmachine test this please

DaveCTurner commented on server/src/main/java/org/elasticsearch/common/network/NetworkAddress.java:
```diff
@@ -88,21 +90,64 @@ public static String format(InetSocketAddress address) {
+     * @param portsRange range of ports
+     * @return formatted string
+     */
+    public static String format(InetAddress address, PortsRange portsRange) {
```
Please would you add a test case to `NetworkAddressTests` for this new behaviour?

mariaral replied:
Ok, I pushed another commit with the relative tests.

DaveCTurner reviewed (commented):
I think we're missing a test case (see comment) but otherwise this looks good.

DaveCTurner commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into feature-bind-host

DaveCTurner commented:
@elasticmachine ok to test

mariaral committed:
Test NetworkAddress.format() with PortsRange

DaveCTurner commented:
Failures look unrelated. @elasticmachine please test this

DaveCTurner commented:
Failures look unrelated. @elasticmachine test this please

DaveCTurner commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into feature-bind-host

DaveCTurner reviewed (approved):
LGTM, but I don't have time to merge this right now.

DaveCTurner commented:
This is merged and backported to 7.x now. Thanks for your contribution @mariaral.
